### ISSUE TITLE
Separate opinions from ordering logic in semver.org spec

### ISSUE THREAD

**Issue by isaacs** [semver/semver#100]
There are numerous places in the semver.org spec where it discusses whether or not a version requirement range is "satisfied by" a specific version.  However, the opinions of version satisfaction vary widely between different platforms.  (Consider Debian version logic, where you specify `X.y.z` and might get `X.y.z-rc.2+build-20130103T124356`, vs npm version logic, where `X.y.z` can only ever match exactly `X.y.z`, but people use comparators and ranges like `>=`, `~>`, `2.3.x`, etc.)

Also, the specification expresses a lot of opinions for software authors, about when they SHOULD, SHOULD NOT, MUST, MUST NOT make certain changes to their version number, based on the changes to their software.  While these opinions are sensible, they are somewhat subjective (eg, what sorts of changes are considered "breaking"), and make it harder for those of us writing semver parsing/comparison implementations to get at what we really care about.

In all of these cases, regardless of anyone's opinion, it's important to know very clearly which versions are greater than, less than, or equal to which other versions.

It would be very helpful if the spec broke just that logic out into a completely separate section, with explicit details, examples, and perhaps even a set of test cases.  I have many for node-semver already, but they diverge from this spec somewhat – a problem I'm trying to correct.  I sometimes get bug reports and pull requests for node-semver that complain about not supporting the latest stuff in the semver 2.0 rc, but there's a lot of language to wade through to find the bits I care about, and some aspects seem unclear.

Specifically, points 10 and 11 discuss this somewhat, but I'm unsure about how to deal with cases where there are multiple numeric parts in the pre-release version.  For example, `1.2.3-a.2.b.55` vs `1.2.3-a.3.b.54`.

Just to be clear: the current spec as it stands is a _very_ helpful guide for software authors, and the opinions it expresses are mostly very sane and reasonable. It's just that the logic for semver package management implementors is somewhat vague.

**Comment by eins78** [comment]
@nichtich <https://github.com/cloudflare/semver_bash>

**Comment by bperrybap** [comment]
What wasn't addressed in this thread yet was how to deal with the version strings created from git describe. (or versions that are post a given release vs before a given release).
From earlier:
>What I'm also finding is that there are some versions in practice in npm that semver 2.0 rc calls "invalid", >such as 1.1.1-2-deadbeef. This is the sort of thing output by git describe --tags, which is used to generate >version identifiers by some build tools. In this case, it's sort of like 1.1.2-but-not-released-yet, so it needs >to be > 1.1.1. To do this, we treat M.m.p-<number>-prereleasetag as if the <number> is a 4th numeric >entry in the tuple. Probably this should be another issue, and go into semver 2.1 or something :)

This brings up an interesting point.
So far in the discussions and in the spec there is a notion of "pre-release" versions. In the case above there is a notion of "post release" development versions.
In other words so far semver seems to only be concerned with production releases and the lead up to that that known production release number. So the version extensions / meta data are related to the known upcoming version number.

And for ordering it defines that  X.Y.Z is always > X.Y.Z-blah

In the case of development, there can be a need to have revisions that are post a production release but yet the actual next production release or number is not yet decided.
So when using git and using tags, if you use git describe to automatically create information for your version number, you potentially end up with something like what was above: 1.1.1-2-deadbeef where 1.1.1 was the last release and according to the rules 1.1.1-2-deadbeef would be considered earlier than 1.1.1

Ignoring the git specifics and a specific format of that version string,
(it could be reformatted using dashes and dots to be con format) it seems that under the current spec there is no way to automatically create post release versions during development without actually bumping the real version number.
This seems very unfortunate in that there is so much extensible information after the Z field and yet while there is way to specify and order releases that are leading up to a known X.Y.Z  release, there is no way to indicate or order development releases that are post X.Y.Z but are not necessarily leading up to any known release number.

**Comment by eins78** [comment]
I did some thinking and experimenting a few weeks back regarding this issue. I haven't had time to clean up the code, so I haven't told anyone about it, but since @bperrybap brings it up…

I've built a [small prototype](https://github.com/eins78/git-describe-semver) that "parses" the output from `git semver` and "always" gives you a correct semver. It is of course highly specific to my taste and needs, but I needed to get a feel for it by coding something.
I came to the same conclusion as @bperrybap in that there always needs to be in bump. In fact, this is one of the few arguments of my toy CLI ("*either:* `--bump <major|minor|patch> *or:* `--change <breaking|feature|bugfix>`").
But: I don't this is a "flaw" in the semver spec that needs to be fixed.

In the general case, we could not automatically decide between "pre release-" and "post release dev versions" anyway, because that is a decision made by the humans. In my mind, it is just a property of the branch you're currently in: If the project starts a new major version, then a new branch is created for it and some preparation is done. Regarding semver, the minimal work I'd like to do is setting `--bump=major`   somewhere (in my case, in a CI script). I've chosen `patch` as a default, because that is what I think happens once you merge the first commit after a tagged release (to a designated branch): you are pre-releasing the next path version.

In a scenario where the release process is completely automated (à la [grunt-semantic-release](https://github.com/boennemann/grunt-semantic-release)) I personally would just rely on parsing all the tags from git to find the last version, but this could also be put in build metadata if a project wants to have this.

**Comment by bhetland** [comment]
@eins78 IIRC the "post-release" concept was formerly proposed in semver using the + delimiter instead of -, so e.g. we could have 1.2-beta but 1.2+dev. The + was at some point abandoned due to some incompatibility I think, but was there ever decided on a usable substitute for it? In our case the situation is similar to yours regarding the version numbers are only decided by humans. It seems the difference is that for a "formal" release build we explicitly control the version label, including any beta releases.

However, all other builds are just deemed "modified" w.r.t. to the current version number from the code repository, so they are just (by default) marked with our equivalent of 1.2+modified. We could conceivably also include info like a repo node#, but did not bother too much with this because the build is likely to be based on some locally modified code compared to the code from the repo. Combined with the build date this was deemed sufficient in our case, as the main idea is to identify the binary as something very informal, and a rough approximation of where in the code history it came from. I was hoping to carry on this idea pretty straightforward if/when transitioning to a semver scheme.

The other major concern we are discussing with adopting a new versioning scheme, would be to have a clear distinction of source branching reflected in the version label. At times a fairly large number of branches may be developed in parallel, so even if they all are spun off the same "base version" it is useful to distinguish between them. Sharing of concrete experiences with this would be appreciated!

**Comment by dwijnand** [comment]
`+` clashes with build metadata. I proposed `~` (tilde) I while back: https://github.com/mojombo/semver/issues/124#issuecomment-20484470

I also explored different ways to do handle post-releases without adding a "post-release" concept to semver in https://github.com/mojombo/semver/issues/124#issuecomment-20505330

**Comment by FichteFoll** [comment]
Other related discussion(s): #/200 and the issues that reference it (at the bottom)
My personal stance on this is the following: https://github.com/mojombo/semver/issues/200#issuecomment-47330860 (post-release identifier is introduced and takes over `+`, metadata gets `~`)

Btw, this appears to be a very frequent issue of the current semver spec, as well as the "generation" thing and "denoting the dependency's version".

**Comment by bperrybap** [comment]
I saw that.
In my case, I use git and tags are mainly only used for code that is externally released but internal builds are done all the time during development and testing. There are revisions numbers that must be created for each build that get automatically generated from git information which uses the most recent tag. 

I think it could go a long way for ease of use and adoption if semvar could directly use the output from git describe, when users follow some simple rules when creating their tag names.

The problem with the current way semvar versioning currently works is that it takes a one sided view that version numbers are only for outside consumption for a given release or leading up to a given known release and falls pretty flat when it comes to internal/development builds. 

This makes it a hassle to deal with since you have to play artificial games with version numbers and or create unneeded tags just to create a semver compliant version string that will never see the light of day externally.

**Comment by jwdonahue** [comment]
@isaacs, it appears as if at least some of your concerns were addressed in SemVer 2.0.0. Please close this issue at your earliest possible convenience.
